pyomo/contrib/pynumero/interfaces/cyipopt_interface.py

```diff
@@ -309,6 +310,49 @@ def __init__(self, nlp, intermediate_callback=None, halt_on_evaluation_error=Non
         # cyipopt.Problem.__init__
         super(CyIpoptNLP, self).__init__()
 
+        # Pre-Pyomo 6.7.4.dev0, we had no way to pass the cyipopt.Problem object
+        # to the user in an intermediate callback. This prevented them from calling
+        # the useful get_current_iterate and get_current_violations methods. Now,
+        # we support this by adding the Problem object to the args we pass to a user's
+        # callback. To preserve backwards compatibility, we inspect the user's
+        # callback to infer whether they want this argument. To preserve backwards
+        # compatibility if the user asked for variable-length *args, we do not pass
+        # the Problem object as an argument in this case.
```

Comment:
I am a big fan of backwards compatibility, but in this case, I think I disagree: if the user defined the callback using `*args`, then it is their responsibility to track any changes to our callback API.

Reply:
The counterpoint is that they may reasonably expect the callback API to be stable. Personally, I would rather we didn't support `*args` at all. Maybe we should raise a deprecation warning if a 12-arg callback or `*args` is provided?

Reply:
The counter-counterpoint is this is in `contrib`, so this is where we make the weakest (i.e., *no*) guarantee on backwards compatibility.

With so many arguments, I like the model where we pass everything by name, and deprecate all use of positional arguments.  We could be clever and even allow callbacks with subsets of named arguments (which would support backwards compatibility), future-proof us to passing new arguments, and remove the current reliance on a large set of ordered arguments.

Reply:
While `CyIpoptInterface` is in contrib, a user can provide a callback via `pyo.SolverFactory("cyipopt", intermediate_callback=callback)`. I've always had the impression that solvers accessible via `SolverFactory` (without some prefix e.g. `"contrib.cyipopt"`), should remain stable. To me, it's a gray area. That said, I do like the idea of only supporting named arguments.

Reply:
@Robbybp how do you want to proceed with this? Should we merge it as-is to get it into the August release and open an issue to deprecate positional arguments in favor of named arguments? Or wait and modify this PR directly?

Reply:
Let's merge this as-is to get it in for the release. I created issue #3354 to track this discussion.